src/print.S

```diff
@@ -62,57 +270,506 @@ print_int:
     imul %ebx, %r10d
     subq %r10, %r9
 
+    incl %r12d
     cmpq $0, %r8
-    ja .Lprint_int_loop
-
+    ja .Lprint_number_loop
+    cmpl %r13d, %r14d
+    jge .Lprint_number_return
+    movl %r14d, %edi
+    subl %r13d, %edi
+    callq append_zeros
+.Lprint_number_return:
+    movl $0x0A, %eax
+    movl %eax, -4(%rbp)
+    leaq -4(%rbp), %rdi
+    callq print_char
     nop
-    popq %r10
-    popq %r9
-    popq %r8
+    popq %r13
+    popq %r12
+    popq %rbx
     movq %rbp, %rsp
     popq %rbp
     retq
 
 /* ======================================================
- * print a single character using the write syscall
+ * Split the IEEE754 floating-point number into it's sub parts:
+ * Sign bit, exponent and mantissa.
+ *
  * parameter:
- *  rdi => pointer to character
+ *  rdi => pointer to structure:
+ *      +0 => sign bit
+ *      +4 => exponent
+ *      +8 => mantissa
+ *
+ * parameter:
+ *  xmm0 => input floating-point number
  */
-print_char:
-    movq %rdi, %rsi /* msg argument of write syscall */
-    movq $0x01, %rax /* write syscall */
-    movq $0x01, %rdi /* write to stdout */
-    movq $0x01, %rdx /* length to write (1) */
-    syscall
+split_into_parts:
+    pushq %rbp
+    movq %rsp, %rbp
+    vmovss %xmm0, -8(%rbp)
+    leaq -8(%rbp), %rax
+    movl (%rax), %eax
+    /* 1. sign bit */
+    sarl $31, %eax
+    movzbl %al, %eax
+    movl %eax, (%rdi)
+    /* 2. exponent */
+    leaq -8(%rbp), %rax
+    movl (%rax), %eax
+    /* mantissa bits = 23*/
+    sarl $23, %eax
+    /* (1 << exponent bits(8)) - 1 = 255*/
+    andl $255, %eax
+    movl %eax, 4(%rdi)
+    /* 3. mantissa */
+    leaq -8(%rbp), %rax
+    movl (%rax), %eax
+    movl .LMantissa_mask, %ebx
+    andl %ebx, %eax
+    movl %eax, 8(%rdi)
+    movq %rbp, %rsp
+    popq %rbp
     retq
+.LMantissa_mask:
+    /* (1 << mantissa bits(23)) -1 */
+    .long 8388607
 
 /* ======================================================
- * print a '=' using the write syscall
+ * Check if the number is equal to zero.
+ *
+ * parameter:
+ *  rdi => pointer to structure:
+ *      +0 => sign bit
+ *      +4 => exponent
+ *      +8 => mantissa
+ *
+ * return:
+ *  rax => 1 (True is zero) | 0 (False is not zero)
  */
-print_equal_sign:
-    pushq %r8
-    movl $0x3D, %r8d
-    leaq (%r8), %rdi
-    callq print_char
-    popq %r8
+zero_check:
```

Comment:
Note that, because of the IEEE 754 formatting for zero, this can be implemented with a couple instructions on the result of `vmovss` (let's imagine it's in %eax, as above):

```asm
zero_check:
    andl $7FFFFFFF, %eax  # Drop sign bit.
    setz %al
    andl $1, %eax
    ret
````

(Personally, it's so short that I would have done the `vmovss` instruction in main, checked for 0 there, and then if it wasn't zero, _then_ pass the floating-point number to split_into_parts, via %rsi [the second argument to a function].)